datafusion/physical-expr-common/src/physical_expr.rs

```diff
@@ -283,6 +284,51 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
     /// See the [`fmt_sql`] function for an example of printing `PhysicalExpr`s as SQL.
     ///
     fn fmt_sql(&self, f: &mut Formatter<'_>) -> fmt::Result;
+
+    /// Take a snapshot of this `PhysicalExpr` if it is dynamic.
+    /// This is used to capture the current state of `PhysicalExpr`s that may contain
+    /// dynamic references to other operators in order to serialize it over the wire
+    /// or treat it via downcast matching.
+    ///
+    /// You should not call this method directly as it does not handle recursion.
+    /// Instead use [`snapshot_physical_expr`] to handle recursion and capture the
+    /// full state of the `PhysicalExpr`.
+    ///
+    /// This is expected to return "simple" expressions that do not have mutable state
+    /// and are composed of DataFusion's built-in `PhysicalExpr` implementations.
+    /// Callers however should *not* assume anything about the returned expressions
+    /// since callers and implementers may not agree on what "simple" or "built-in"
+    /// means.
+    /// In other words, if you need to searlize a `PhysicalExpr` across the wire
+    /// you should call this method and then try to serialize the result,
+    /// but you should handle unknown or unexpected `PhysicalExpr` implementations gracefully
+    /// just as if you had not called this method at all.
+    ///
+    /// In particular, consider:
+    /// * A `PhysicalExpr` that references the current state of a `datafusion::physical_plan::TopK`
+    ///   that is involved in a query with `SELECT * FROM t1 ORDER BY a LIMIT 10`.
+    ///   This function may return something like `a >= 12`.
+    /// * A `PhysicalExpr` that references the current state of a `datafusion::physical_plan::joins::HashJoinExec`
+    ///   from a query such as `SELECT * FROM t1 JOIN t2 ON t1.a = t2.b`.
+    ///   This function may return something like `t2.b IN (1, 5, 7)`.
+    ///
+    /// A system or function that can only deal with a hardcoded set of `PhysicalExpr` implementations
+    /// or needs to serialize this state to bytes may not be able to handle these dynamic references.
+    /// In such cases, we should return a simplified version of the `PhysicalExpr` that does not
+    /// contain these dynamic references.
+    ///
+    /// Systems that implement remote execution of plans, e.g. serialize a portion of the query plan
+    /// and send it across the wire to a remote executor may want to call this method after
+    /// every batch on the source side and brodcast / update the current snaphot to the remote executor.
+    ///
+    /// Note for implementers: this method should *not* handle recursion.
+    /// Recursion is handled in [`snapshot_physical_expr`].
+    fn snapshot(&self) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+        // By default, we return None to indicate that this PhysicalExpr does not
+        // have any dynamic references or state.
+        // This is a safe default behavior.
+        Ok(None)
+    }
```

Comment:
I'm wondering if it would be reasonable to inject the schema here so that this can be used for the use case in https://github.com/apache/datafusion/pull/15057 as well: we could model a predicate like `variant_get(col, 'foo') = 5` by making `variant_get(col, 'foo')` a dynamic physicalexpr that snapshots itself to `"col.typed_value.foo.typed_value` if the schema/types permit otherwise it snapshots itself to `_variant_get_value(col.metadata, col.value)`.

Reply:
What do you mean "inject" the schema?

Like make this something like?
```rust
    fn snapshot(&self, &Schema) -> Result<Option<Arc<dyn PhysicalExpr>>> {
```

Reply:
Yes precisely.

So instead of being just dynamic on some internal state it will be dynamic on the actual physical schema of the file it is being applied to. But maybe that should be its own API:

`fn with_physical_schema(&self, schema: &Schema) -> Option<Arc<dyn PhysicalExpr>>`

Then if something wants a snapshot that's specialized for a specific file schema it can call both methods.

Reply:
in theory the schema of the input to a PhysicaExpr shouldn't change so any expr that needs it could hold a reference 🤔

Reply:
Is that true? Where would it hold onto the reference from, `data_type()`? Can't the same `PhysicalExpr` be used in multiple contexts, e.g. with the same columns but different projections / different physical column orders? Put another way, is it guaranteed that `data_type()` can only be called once or that `evaluate()` is always called with the same schema?

To be clear: I don't think this is important for this PR. After reflection in https://github.com/apache/datafusion/pull/15568/files#r2033710338 I think if we want that we can add it as a new method in https://github.com/apache/datafusion/pull/15057

Reply:
> Is that true? Where would it hold onto the reference from, data_type()? Can't the same PhysicalExpr be used in multiple contexts, e.g. with the same columns but different projections / different physical column orders? 

Not in my understanding -- while it could in theory work this way I don't think they are typically shared. 

> Put another way, is it guaranteed that data_type() can only be called once or that evaluate() is always called with the same schema?

In my mind what happens is that the PhysicalExpr is created with the schema information: https://docs.rs/datafusion/latest/datafusion/physical_expr/fn.create_physical_expr.html

But it doesn't hold a reference to the schema internally 🤔